CC: Software Sustainability Matrix (#1833)

bartlettroscoe opened the pull request:
# Description

EB Member: @pagrubel 

Resolves:

* #1833

NOTE: The wikize_refs.py tool worked to process the initial references but then could not reprocess the same file.  So future changes to the references will need to be done manually.

## PR checklist for files displayed on bssw.io site

* [x] ***[Author]*** `@mention` the BSSw.io editorial board member `@<eb-member-id>` in **Description** above assigned to shepherd your PR.
* [x] ***[Author]*** Add the `<issue-id>` in the **Description** above for the associated GitHub Issue.
* [x] ***[Author]*** Ensure `wikize_refs.py -i <base>.md` is run and commit (if using [wikize_refs.py]).
* [x] ***[Author]*** Inspect the content in the `*.md` file(s) as rendered in GitHub for this PR.
* [x] ***[EB Mem]*** Assign this PR to the EB member `<eb-member-id>`.
* [x] ***[EB Mem]*** Assign this PR to the author of the PR `<pr-author-id>`.
* [x] ***[EB Mem]*** Add one or more Reviewers.
* [x] ***[EB Mem]*** Add label `content: <content-type>` for the type of contribution.
* [x] ***[EB Mem]*** Add to Project `Content Development` (see [Content Development]).
* [x] ***[EB Mem]*** Add [meta-data] to the `*.md` file(s) (set `Publish: yes`).
* [x] ***[EB Mem]*** Add label `preview` (so PR branch will be merged to 'preview' branch and watch for possible merge failures).
* [x] ***[EB Mem]*** Rebuild [preview] site and confirm new content is there, renders correctly and is returned in searches.
* [x] ***[Author]*** Make any final changes to the PR based on feedback.
* [x] ***[Author]*** Ensure `wikize_refs.py -i <base>.md` is run and commit (if using [wikize_refs.py]).
* [x] ***[EB Mem]*** Rebuild [preview] site and re-confirm content looks correct.
* [x] ***[EB Mem]*** Ensure at least one reviewer signs off on the final changes.
* [x] ***[EB Mem]*** Move the PR to "Ready to Publish" in [Content Development].
* [ ] ***[EB Mem]*** Leave comment and @mention `@betterscientificsoftware/bssw-maint` (**BSSw Maint**) asking to carry out final publication steps.
* [ ] ***[BSSw Maint]*** Verify that all needed files are present in the PR (article, images, updates to Site/Homepage.md carousel and/or Site/Announcements/Announcements.md as appropriate).
* [ ] ***[BSSw Maint]*** Merge PR. (Should automatically move to "Done" in [Content Development].)
* [ ] ***[BSSw Maint]*** Verify new contribution shows up on [bssw.io] as expected.

NOTE:
* Checklist items prefixed with ***[Author]*** are expected to be performed by the author of the PR or can be performed by the author.
* Checklist items prefixed with ***[EB Mem]*** must be performed by a [BSSw.io Editorial Board (EB) Member](https://betterscientificsoftware.github.io/bssw.io/bssw_members.html).
* Checklist items prefixed with ***[BSSw Maint]*** must be performed by someone in the `@betterscientificsoftware/bssw-maint` team (hint: type `@`, `b`, `s`, `s`, `w`, `-`, `m`  to auto-complete to `@betterscientificsoftware/bssw-maint`).

<!-- Standard links below, leave these this section! -->

[preview]: https://preview.bssw.io
[bssw.io]: https://bssw.io
[Content Development]: https://github.com/betterscientificsoftware/bssw.io/projects/3
[BSSw Internal]: https://github.com/betterscientificsoftware/bssw.io/projects/2
[meta-data]: https://betterscientificsoftware.github.io/bssw.io/bssw_styling_common.html#metadata-section
[wikize_refs.py]: https://github.com/betterscientificsoftware/bssw.io/blob/master/utils/README.md#wikize_refspy

bartlettroscoe commented:
FYI: I rebuilt the preview site and you can see the generated page here:

* https://preview.bssw.io/items/software-sustainability-matrix

For some reason, the figure is not being shown.  It could be that I tried to shrink the size of the figure and some of those `img` options may not be supported.

bartlettroscoe commented:
After pushing the last commit 2bc0e4fe5bb7b874acbe4e8bd6afc90d8d29f7d7, the page:

* https://preview.bssw.io/items/software-sustainability-matrix

now displays the figure but it is fuzzy.

bartlettroscoe commented:
FYI: My latest commit 8323ff2c6be310a4b4473cceb16f3d6119c95fcf replaced the figure with one directly downloaded from the source article and now it is not fuzzy at:

* https://preview.bssw.io/items/software-sustainability-matrix

but now it is very tiny showing:

----

![image](https://github.com/betterscientificsoftware/bssw.io/assets/1959736/9c7c435c-cfc4-4160-b068-ef94f788dffc)

----

But GitHub renders this just fine at:

* https://github.com/bartlettroscoe/bssw.io/blob/1833-kitware-ssm/CuratedContent/SoftwareSustainabilityMatrix.md

showing:

----

![image](https://github.com/betterscientificsoftware/bssw.io/assets/1959736/1ba9b6ff-21b7-479c-b7e3-3571327c702f)

----

so I don't understand what the bssw.io site generator is doing to these figures.

markcmiller86 commented:
@bartlettroscoe I used Safari's **Develop** menu to **Show page source** to debug things a bit.

That image is getting handled by this HTML code in the generated site...

```
<img src="https://res.cloudinary.com/bssw/image/fetch/w_200,h_200,c_fit/https://raw.githubusercontent.com/betterscientificsoftware/bssw.io/preview/images/SoftwareSustainabilityMatrix.png?raw=true" class="logo">
```

Which appears as this ...

<img src="https://res.cloudinary.com/bssw/image/fetch/w_200,h_200,c_fit/https://raw.githubusercontent.com/betterscientificsoftware/bssw.io/preview/images/SoftwareSustainabilityMatrix.png?raw=true" class="logo">

Note the `w_200,h_200,c_fit` in the image path.

However, if you adjust that ref a bit, removing the `w_200,h_200,c_fit`, you get a better image...

```
<img src="https://res.cloudinary.com/bssw/image/fetch/https://raw.githubusercontent.com/betterscientificsoftware/bssw.io/preview/images/SoftwareSustainabilityMatrix.png?raw=true" class="page">
```

Which appears as this...

<img src="https://res.cloudinary.com/bssw/image/fetch/https://raw.githubusercontent.com/betterscientificsoftware/bssw.io/preview/images/SoftwareSustainabilityMatrix.png?raw=true" class="page">

I recall the backend having trouble with images like this in other contexts and having to do something *special* to make it work (a light box is added in the generated text).

So, I think the solution is to reference the image like this...

```
<img src='../../images/SoftwareSustainabilityMatrix.png' class='page' />[Figure 1: Software Sustainability Matrix]
```

markcmiller86 commented:
That is, for example, what is done in [this article](https://github.com/betterscientificsoftware/bssw.io/blob/main/Articles/Blog/2020-11-PSIP4HDF5.md)

markcmiller86 commented:
The `class=logo` attribute in the `<img>` tag also had me a bit worried too. I worry that a *logo* image may get resized to something smaller. So, I would not include that.

In the proposed fix, `class=page` I think offers the opportunity for the image to be *clickable* and then bring up a lightbox with an enlarged image. That isn't really my favored way to handle it. I'd just like it to load the raw, true sized image. But, the generated site wasn't going to handle that well.

bartlettroscoe commented:
> The `class=logo` attribute in the `<img>` tag also had me a bit worried too. I worry that a _logo_ image may get resized to something smaller. So, I would not include that.

@markcmiller86, thanks for the tip!  I made that change in commit f9d9ec4aed32e21ad5b888174db000be12cb1d8d and rebulit the page:

* https://preview.bssw.io/items/software-sustainability-matrix

and now it fills the entire page:

----

![image](https://github.com/betterscientificsoftware/bssw.io/assets/1959736/116b4073-8b43-4c71-b8a4-24c91223785b)

----

NOTE: The official documentation at:

* https://betterscientificsoftware.github.io/bssw.io/bssw_styling_common.html#body-images

says to use `class='logo'` at:

```
<br>
<img src='../../images/YOUR-IMAGE-NAME.png' class='logo'/>[optional caption]
<br>
```

But then I noticed reading down it does mention the `page` class as an option that does seem to meet my desired usage.

But I wonder if the default image class in the example at:

* https://betterscientificsoftware.github.io/bssw.io/bssw_styling_common.html#body-images

might be changed from `logo` to `page`?

But it does seem like the bssw.io site generator is a little more finicky abouts the display of images compared to other rendering tools for Markdown (like GitHub and the Chrome Markdown Viewer Extension).  (With those Markdown renderers, I was able to have more control over how the image was displayed like at 80% of the page width, for example.)

bartlettroscoe commented:
Hello @pagrubel, I think this is ready for final review. Take a look at the rendered page:
 
* https://preview.bssw.io/items/software-sustainability-matrix

bartlettroscoe commented:
Hello @pagrubel, did you get this notification?

pagrubel commented on CuratedContent/SoftwareSustainabilityMatrix.md:
```diff
@@ -0,0 +1,81 @@
+
+The Software Sustainability Matrix (SSM) is a joint effort of the Scientific Computing and Imaging (SCI) Institute and Kitware to improve the health of the open-source package ecosystem making up the foundational software portfolio for the Center for Integrative Biomedical Computing (CIBC) lead by the SCI for the National Institutes of Health (NIH) <sup>[1],[2]</sup>.
+The goal of the SSM effort is to create numerical scores for different aspects of software sustainability for a software package as well of a combined sustainability score.
+These SSM sustainability scores are being to guide efforts to improve the sustainability of the software package ecosystem making to CIBC software portfolio, with the hopes that it will encourage support and contributions from the open-source community and reduce the cost and risks to sustain this collection software.
```
@bartlettroscoe I tried to make a suggestion to fix this sentence starting with changing "being" to "being used", but then stumbled on what you are saying by "making to CIBC software porfolio". I think you will need to fix it.

bartlettroscoe replied:
@pagrubel, I reworded this sentence in commit e325da88e801036e9a5daf42009a94239c9f5437.  Please take a look.

pagrubel commented on CuratedContent/SoftwareSustainabilityMatrix.md:
```diff
@@ -0,0 +1,81 @@
+
+to produce a single SSM Score [0-100], where F<sup>I</sup> + F<sup>R</sup> + F<sup>C</sup> + F<sup>T</sup> = 1.
+The SCI & Kitware process<sup>[3]</sup> typically uses the weighting factors F<sup>I</sup> = F<sup>C</sup> = 1/3, and F<sup>R</sup> = F<sup>T</sup> = 1/6.
+It is not specified how the SCI & Ktiware program computes the component scores in the four areas 'I', 'R', 'C', and 'T' for the software packages of interest and the authors state that computing these individual component scores is ultimately a subjective matter.
```
```suggestion
It is not specified how the SCI & Kitware program computes the component scores in the four areas 'I', 'R', 'C', and 'T' for the software packages of interest, and the authors state that computing these individual component scores is ultimately a subjective matter.
```

bartlettroscoe replied:
I applied the suggested change in commit e325da88e801036e9a5daf42009a94239c9f5437.

pagrubel commented on CuratedContent/SoftwareSustainabilityMatrix.md:
```diff
@@ -0,0 +1,81 @@
+In this respect, this is similar to the impact of the OpenSSF Best Practices Badge Program<sup>[4]</sup>.
+That is, by having a team fill out each best practice item, the team will often implement suggested best practices, especially for practices that are not too difficult to adopt and support.
+
+In summary, the Software Sustainability Matrix (SSM) is a promising approach to help quantify the sustainability of individual software packages (and collectively entire software package ecosystems) and then help to drive efforts to make targeted improvements to improve sustainability.
```
```suggestion
In summary, the Software Sustainability Matrix (SSM) is a promising approach to help quantify the sustainability of individual software packages (and collectively entire software package ecosystems) and subsequently help to drive efforts to make targeted improvements for sustainability.
```

bartlettroscoe replied:
I applied the suggested change in commit e325da88e801036e9a5daf42009a94239c9f5437.

pagrubel reviewed (changes requested):
@bartlettroscoe I've made a few suggestions. I have not looked over the references for formatting. I'll do that in the next iteration. The article looks good and inspires me to start using the matrix.

pagrubel commented:
@bartlettroscoe I think there are still a few changes to be made

bartlettroscoe commented:
> @bartlettroscoe I think there are still a few changes to be made

Hello @pagrubel, I pushed the commit e325da88e801036e9a5daf42009a94239c9f5437 last week that should have addressed your review comments [above](https://github.com/betterscientificsoftware/bssw.io/pull/1844#pullrequestreview-1688388714).  Where there other specific issues to be addressed?

bartlettroscoe commented on CuratedContent/SoftwareSustainabilityMatrix.md:
```diff
@@ -14,9 +14,9 @@ Authors | Will Schroeder, Jess Tate, Jean-Christophe Fillion-Robin, Christopher
+The Software Sustainability Matrix (SSM) is a joint effort of the Scientific Computing and Imaging (SCI) Institute and Kitware to improve the sustainability of the open-source package ecosystem making up the foundational software portfolio for the Center for Integrative Biomedical Computing (CIBC) lead by the SCI for the National Institutes of Health (NIH) <sup>[1],[2]</sup>.
 The goal of the SSM effort is to create numerical scores for different aspects of software sustainability for a software package as well of a combined sustainability score.
-These SSM sustainability scores are being to guide efforts to improve the sustainability of the software package ecosystem making to CIBC software portfolio, with the hopes that it will encourage support and contributions from the open-source community and reduce the cost and risks to sustain this collection software.
+This SSM process is being used to guide efforts to improve the sustainability of the CIBC software package ecosystem, with the hopes to transition the continuing support of this software to the open-source community<sup>[1]</sup>.
```
Addressed in new commit new commit 071b872eebba57400c7233f0665be53bb7e2f6db.

bartlettroscoe commented on CuratedContent/SoftwareSustainabilityMatrix.md:
```diff
@@ -42,14 +42,14 @@ Scores in the range [0-100] in these four areas Impact, Risks, Community, and Te
 
 It is stated by the authors that the primary benefit of estimating the SSM score by a software package team is not in computing the final score, but instead in highlighting areas of possible improvement.
-It is stated that many teams make significant improvements in the sustainability for their packages while working through in the process of estimating the SSM score by adopting better processes in different areas.
+It is claimed that many teams make significant improvements in the sustainability for their packages while working through in the process of estimating the SSM score by adopting better processes in different areas.
```
Addressed in new commit new commit 071b872eebba57400c7233f0665be53bb7e2f6db.

bartlettroscoe commented on CuratedContent/SoftwareSustainabilityMatrix.md:
```diff
@@ -42,14 +42,14 @@ Scores in the range [0-100] in these four areas Impact, Risks, Community, and Te
+That is, by having a team score each suggested practice, the team will often implement many suggested best practices, especially for practices that are not too difficult to adopt and support.
 
-In summary, the Software Sustainability Matrix (SSM) is a promising approach to help quantify the sustainability of individual software packages (and collectively entire software package ecosystems) and then help to drive efforts to make targeted improvements to improve sustainability.
+In summary, the Software Sustainability Matrix (SSM) is a promising approach to help quantify the sustainability of individual software packages (and collectively entire software package ecosystems) and subsequently to drive efforts to make targeted improvements for sustainability.
```
Addressed in new commit new commit 071b872eebba57400c7233f0665be53bb7e2f6db.
